File: solutions/golang/410/solution.go
var maxNumSubArrays int

func splitArray(nums []int, k int) int {
    maxNumSubArrays = k
    left := math.MaxInt
    right := 0
    for _, num := range nums {
        if (left > num) {
            left = num
        }
        right += num
    }
    
    for left <= right {
        mid := left + (right - left) / 2
        if check(mid, nums) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    
    return left
}

func check(maxSum int, nums []int) bool {
    numSubArrays := 1
    curr := 0
    for _, num := range nums {
        curr += num
        if curr > maxSum {
            curr = num
            numSubArrays++
            if (numSubArrays > maxNumSubArrays) || (curr > maxSum) {
                return false
            }
        }
    }
    
    return true
}

File: solutions/cpp/410/solution.cpp
class Solution {
public:
    int k;
    int splitArray(vector<int>& nums, int k) {
        this->k = k;
        int left = *min_element(nums.begin(), nums.end());
        int right = accumulate(nums.begin(), nums.end(), 0);
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (check(mid, nums)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    bool check(int maxSum, vector<int> &nums) {
        int numSubArrays = 1;
        int curr = 0;
        for (int num : nums) {
            curr += num;
            if (curr > maxSum) {
                numSubArrays++;
                curr = num;
                if ((numSubArrays > k) || (curr > maxSum)) {
                    return false;
                }
            }
        }

        return true;
    }
};
